node/rpc/src/rpc_ws_handler.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_std::sync::{Arc, Mutex};
+use crossbeam::atomic::AtomicCell;
+use futures::{SinkExt, StreamExt};
+use log::{debug, error, info, warn};
+use tide_websockets::{Message, WebSocketConnection};
+
+use beacon::Beacon;
+use blockstore::BlockStore;
+use wallet::KeyStore;
+
+use crate::data_types::{JsonRpcServerState, StreamingData, SubscriptionHeadChange};
+use crate::rpc_util::{
+    call_rpc, call_rpc_str, get_error_str, RPC_METHOD_CHAIN_HEAD_SUB, RPC_METHOD_CHAIN_NOTIFY,
+    RPC_METHOD_CHAIN_NOTIFY_RESPONSE,
+};
+
+pub async fn rpc_ws_handler<DB, KS, B>(
+    request: tide::Request<JsonRpcServerState>,
+    ws_stream: WebSocketConnection,
+) -> Result<(), tide::Error>
+where
+    DB: BlockStore + Send + Sync + 'static,
+    KS: KeyStore + Send + Sync + 'static,
+    B: Beacon + Send + Sync + 'static,
+{
+    let rpc_server = request.state();
+    let (ws_sender, mut ws_receiver) = ws_stream.split();
+    let ws_sender = Arc::new(Mutex::new(ws_sender));
+    let socket_active = Arc::new(AtomicCell::new(true));
+
+    info!("Accepted WS connection!");
+
+    while let Some(message_result) = ws_receiver.next().await {
+        match message_result {
```

Comment:
Something is blocking in this while loop that is preventing me from spamming ourselves with ChainNotify calls. I would consider throwing this whole block into a task. There are certain Websocket calls like StateMessageWait that can take up a few minutes before a response (as it waits X epochs for transaction confirmations). And this will prevent any new requests from going through. I think this is why I cant spam us with ChainNotify calls.

Reply:
That makes sense, I'll do that in the morning.